cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -0,0 +1,194 @@
+	const (
+		short = "Generate initramfs mount tuples"
+		long  = "Generate mount tuples for the initramfs until nothing more can be done"
+	)
+
+	if _, err := parser.AddCommand("initramfs-mounts", short, long, &cmdInitramfsMounts{}); err != nil {
+		panic(err)
+	}
+}
+
+type cmdInitramfsMounts struct{}
+
+func (c *cmdInitramfsMounts) Execute(args []string) error {
+	return generateInitramfsMounts()
+}
+
+var (
+	// the kernel commandline - can be overridden in tests
+	procCmdline = "/proc/cmdline"
+
+	// Stdout - can be overridden in tests
+	stdout io.Writer = os.Stdout
+)
+
+var (
+	runMnt = "/run/mnt"
+
+	osutilIsMounted = osutil.IsMounted
+)
+
+// generateMountsMode* is called multiple times from initramfs until it
+// no longer generates more mount points and just returns an empty output.
+func generateMountsModeInstall() error {
+	// 1. always ensure seed partition is mounted
+	isMounted, err := osutilIsMounted(filepath.Join(runMnt, "ubuntu-seed"))
+	if err != nil {
+		return err
+	}
+	if !isMounted {
+		fmt.Fprintf(stdout, "/dev/disk/by-label/ubuntu-seed %s\n", filepath.Join(runMnt, "ubuntu-seed"))
+		return nil
+	}
+	// XXX: how do we select a different recover system from the cmdline?
+
+	// 2. (auto) select recovery system for now
+	isMounted, err = osutilIsMounted(filepath.Join(runMnt, "base"))
+	if err != nil {
+		return err
+	}
+	if !isMounted {
+		l, err := filepath.Glob(filepath.Join(runMnt, "/ubuntu-seed/systems/*"))
+		if err != nil {
+			return err
+		}
+		if len(l) == 0 {
+			return fmt.Errorf("cannot find a recovery system")
+		}
+		if len(l) > 1 {
+			return fmt.Errorf("cannot use multiple recovery systems yet")
+		}
+		// load the seed and generate mounts for kernel/base
+		label := filepath.Base(l[0])
+		seedDir := filepath.Dir(filepath.Dir(l[0]))
+		deviceSeed, err := seed.Open(seedDir, label)
+		if err != nil {
+			return err
+		}
+		// load assertions into a temporary database
+		if err := deviceSeed.LoadAssertions(nil, nil); err != nil {
+			return err
+		}
+		perf := timings.New(nil)
+		if err := deviceSeed.LoadMeta(perf); err != nil {
+			return err
+		}
+		// XXX: do we need more cross checks here?
+		for _, essentialSnap := range deviceSeed.EssentialSnaps() {
+			snapf, err := snap.Open(essentialSnap.Path)
+			if err != nil {
+				return err
+			}
+			info, err := snap.ReadInfoFromSnapFile(snapf, essentialSnap.SideInfo)
+			if err != nil {
+				return err
+			}
+			switch info.GetType() {
+			case snap.TypeBase:
+				fmt.Fprintf(stdout, "%s %s\n", essentialSnap.Path, filepath.Join(runMnt, "base"))
+			case snap.TypeKernel:
+				fmt.Fprintf(stdout, "%s %s\n", essentialSnap.Path, filepath.Join(runMnt, "kernel"))
+			}
+		}
+
+		return nil
+	}
+
+	// 3. mount "ubuntu-data" on a tmpfs
+	isMounted, err = osutilIsMounted(filepath.Join(runMnt, "ubuntu-data"))
+	if err != nil {
+		return err
+	}
+	if !isMounted {
+		// XXX: is there a better way?
+		fmt.Fprintf(stdout, "--type=tmpfs tmpfs /run/mnt/ubuntu-data\n")
```

Comment:
this is perfect

Reply:
@mvo5 how will initramfs know this is the one to use as / ?  same question around base ?

Reply:
@xnox do you want something more explicit for the boot protocol that tells the initramfs what to map where? I.e. one way (the current) is to just ouput "--type=tmpfs tmpfs /run/mnt/ubuntu-data" and the initramfs will have to know that ubuntu-data is what will be "/". We could be more explicit and add something like "--target" or similar. My gut feeling is that the current approach is the right split between the concerns, i.e. snap-boostrap tells the initramfs "here is base,kernel,writable(ubuntu-data)" and the initramfs puts things into the right place (i.e. constructs "/" based on base and puts the kernel modules into the right sub directory etc). WDYT?

Reply:
The target is `/sysroot` which is where we will pivoted.

Reply:
does this mean the target needs to be changed to that?